docs/operational_guide/fileset_migrations.md

```diff
@@ -0,0 +1,41 @@
+# Fileset Migrations
+
+Occasionally, changes will be made to the format of fileset files on disk. When those changes need to be applied to already existing filesets, a fileset migration is required. Migrating existing filesets is beneficial so that improvements made in newer releases can be applied to all filesets, not just newly created ones.
+
+## Migration Process
+Migrations are executed during the initial stages of the bootstrap. When enabled, the filesystem bootstrapper will scan for filesets that should be migrated and migrate any filesets found. A fileset is determined to be in need of a migration based on the `MajorVersion` and `MinorVersion` found in the info file. If `MajorVersion.MinorVersion` is less than the target migration version, then that fileset will be scheduled for migration.
+
+If migrations are deemed necessary, the bootstrap process pauses until the migrations complete. If a failure occurs while migrating, an error is logged and the process continues. If a fileset is not successfully migrated, the non-migrated version of the fileset is used going forward. In other words, whether they succeed or fail, migrations should leave filesets in a good state.
```

Comment:
it seems unfortunate that migrations slow down the bootstrapping process, especially since it seems like the migration isn't required for boot up (i.e if it fails, the old fs is used). curious if there is any future plan to change this.

Reply:
The concurrency knobs help a bit to make them happen pretty quickly, but, yeah, not ideal for sure. This is something we can revisit if it starts to become a problem and we have more migrations. Given that m3 has been around for a while now and we're just implementing something like this, I'm inclined to wait and see before investing more time in optimizations.